gbpservice/neutron/nsf/proxy/proxy.py

```diff
@@ -0,0 +1,223 @@
+        self.rest_server_address = config.get('OPTIONS', 'rest_server_address')
+        self.rest_server_port = config.getint('OPTIONS', 'rest_server_port')
+        self.max_connections = config.getint('OPTIONS', 'max_connections')
+        self.worker_threads = config.getint('OPTIONS', 'worker_threads')
+        self.connect_max_wait_timeout = config.getfloat(
+            'OPTIONS', 'connect_max_wait_timeout')
+        self.idle_max_wait_timeout = config.getfloat(
+            'OPTIONS', 'idle_max_wait_timeout')
+        self.idle_min_wait_timeout = config.getfloat(
+            'OPTIONS', 'idle_min_wait_timeout')
+
+
+class UnixServer(object):
+
+    def __init__(self, conf, proxy):
+        self.proxy = proxy
+        self.bind_path = conf.unix_bind_path
+        self.max_connections = conf.max_connections
+        # Make sure the socket does not already exist
+        try:
+            os.unlink(self.bind_path)
+        except OSError:
+            if os.path.exists(self.bind_path):
+                raise
+
+        # Create a UDS socket
+        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+
+        # Bind the socket to the port
+        print >>sys.stderr, 'starting up on %s' % self.bind_path
+        self.socket.bind(self.bind_path)
+        self.socket.listen(self.max_connections)
+
+    def listen(self):
+        client, address = self.socket.accept()
+        self.proxy.new_client(client, address)
+
+
+class TcpClient(object):
+
+    def __init__(self, conf, proxy):
+        self.conf = conf
+        self.proxy = proxy
+        self.server_address = conf.rest_server_address
+        self.server_port = conf.rest_server_port
+        # Connect the socket to the port where the server is listening
+        self.server = (self.server_address, self.server_port)
+
+    def connect(self):
+        sock = socket.socket()
+        print >>sys.stderr, 'connecting to %s port %s' % self.server
+        sock.settimeout(self.conf.connect_max_wait_timeout)
+        try:
+            sock.connect(self.server)
+        except socket.error, exc:
+            print "Caught exception socket.error : %s" % exc
+            return sock, False
+        return sock, True
+
+
+class Connection(object):
+
+    def __init__(self, conf, socket):
+        self._socket = socket
+        self._idle_wait = conf.idle_min_wait_timeout
+        self._idle_timeout = conf.idle_max_wait_timeout
+        self._idle_count_max = (self._idle_timeout / self._idle_wait)
+        self._idle_count = 0
+
+    def _tick(self):
+        self._idle_count += 1
+
+    def _timedout(self):
+        if self._idle_count > self._idle_count_max:
+            raise ConnectionTimedOut
+
+    def idle(self):
+        self._tick()
+        self._timedout()
+
+    def idle_reset(self):
+        self._idle_count = 0
+
+    def recv(self):
+        self._socket.settimeout(self._idle_wait)
+        try:
+            data = self._socket.recv(16)
```

Comment:
With select also we would need a timeout. For some reason python select is not working like 'c' select. it is blocking the context.